File: XECI_matlab/xeci_corr.m
function [esXECI, ciXECI, tsXECI, varXECI] = xeci_corr(sampleR, nSize, ...
                                                       nPartialR, nullRho, ciSize, displayOut, nDecs)
% 
% XECI_CORR calculates effect sizes and CIs for Pearson zero-order and 
% partial correlations, and CI for Spearman rank-order correlation (if 
% Spearman correlation entered as input value)
%
%
%   INPUT:  
%       sampleR    = observed correlation value (scalar or k X 1 vector)
%       nSize      = sample size (scalar or k X 1 vector)
%       nPartialR  = number of partialled variables (scalar or k X 1 vector)
%       nullRho    = null hypothesised rho value
%       ciSize     = confidence interval size
%       displayOut = display output (1 = yes, 0 = no)
%       nDecs      = number of displayed decimals (default = 4)
%
%
%   OUTPUT:
%       esXECI = (2 X 1) vector of point estimates of effect sizes 
%                    Sample correlation (zero-order or partial)
%                    Olkin & Pratt's unbiased estimate
%
%       ciXECI = (8 X 3) matrix of lower and upper bound CI estimates
%                    Large-sample Pearson
%                    Large-sample Olkin & Pratt's unbiased estimate
%                    Large-sample Spearman 
%                    Exact for Pearson correlation
%                    Exact for unbiased estimate of Pearson
%
%       tsXECI = (3 X 1) vector of t test results
%                    t statistic value
%                    Degrees of freedom
%                    Obtained P value
%                    Null-rho P value
%
%       varXECI = (9 X 2) vector of estimated variances
%                    Large sample variance
%                    Olkin & Pratt variance
%
% Notes: (1) If any non-scalar t, nu, or p are used, then each must be 
%            k X 1 conformable
%        (2) If any scalar {t, nu, p} mixed with k X 1 vector {t, nu, p}, 
%            then k X 1 vector of duplicated scalar values created by 
%            MATLAB DISTCHCK function
%        (3) IF FZERO fails with c1 = -1, then RHO = NaN
%
%
% See also rhocdf, ncp_rho
%

%
% VERSION HISTORY
%     Created:    11 Jul 2008 
%                 13 Mar 2015:   Changed cPearson / cSperman to sePearson / seSperman
%
%


%% Initialise output arguments

    esXECI = [];
    ciXECI = [];
    tsXECI = [];
    varXECI = [];


    %% ----------------------------------------Initial check of input arguments
    if nargin < 5
        error('XECI_corr:IncorrectInputs','Requires at least five input arguments.');
    end

    if (displayOut ~= 1 && displayOut ~= 0)
        error('XECI_corr:IncorrectDPO','Display output flag not equal to 0 or 1.');
    end

    [errorcode, sampleR, nSize, nPartialR] = distchck(3,sampleR,nSize,nPartialR);

    if errorcode > 0
        error('XECI_corr:InputSizeMismatch','Requires non-scalar arguments to match in size.');
    end


    % Generates Dialog Box error messages for input errors

    if (sampleR  <  -1) || (sampleR > 1)
        errordlg( 'Sample correlation value is < -1 or > +1 ', ...
                 'Input Entry Error' ); 
        return

    elseif (nSize  <  1)
        errordlg( 'Sample size is less than or equals 0 ', ...
                 'Input Entry Error' ); 
        return

    elseif (nPartialR  <  0) || (nPartialR > nSize)
        errordlg( 'Number of partialled variables < 0 or > N ', ...
                 'Input Entry Error' ); 
        return

    elseif (nullRho  <  -1) || (nullRho > 1)
        errordlg( 'Null hypothesised correlation value is < -1 or > +1 ', ...
                 'Input Entry Error' ); 
        return
    end

    if (ciSize <= 0 || ciSize >= 100)
        errordlg( 'Confidence interval value < 0 or > 100 ', ...
                 'Input Entry Error' ); 
        return
    end
    
    % Default number of decimal places displayed
    if nargin == 6
       nDecs = 4 ;
    end ;

    % No output displayed
    if nargin == 5
       displayOut = 0 ;
    end ;


    

    %% ------------------------------------------------Preliminary calculations

    nSizeP = nSize - nPartialR;
    nSizeP1 = nSize - nPartialR - 1;

    probCI = [((100+ciSize)/200) ((100-ciSize)/200)];                       % P values for defining CI bounds

    zCritical = norminv(1 - probCI, 0, 1);                                  % Cricial Z statistics values for CI



    %% ------------------------------------------------Effect size calculations

    unbiasedR = sampleR * ...                                               % Olkin & Pratt unbiased estimate of r
                hypergeomF(0.5, 0.5, (nSizeP1-1)/2, 1 - sampleR^2);



    %% ---------------------------------------Confidence intervals calculations

    seSampleR = sqrt(1/(nSizeP - 3));

    ciSampleR  = tanh(atanh(sampleR)   + zCritical * seSampleR );           % Large sample Fisher r-to-z confidence intervals 
    ciUnbiased = tanh(atanh(unbiasedR) + zCritical * seSampleR );           % for sample correlation and unbiased correlation
    
    ciExact = ncp_rho(sampleR, nSizeP, nPartialR, probCI);                  % Exact confidence interval for sample correlation

    ciExactUnb = ncp_rho(unbiasedR, nSizeP, nPartialR, probCI);             % Exact confidence interval for unbiased correlation

    logCorr = 0.5*(log(1 + sampleR) - log(1 - sampleR));

    sePearson = (1 * zCritical) ./ sqrt(nSize - 3);
    seSpearman = (sqrt(1 + sampleR^2/2) .* zCritical) ./ sqrt(nSize - 3);

    pearson_L = logCorr + sePearson;

    spearman_L = logCorr + seSpearman;

    ciPearson = (exp(2*pearson_L) - 1) ./ (exp(2*pearson_L) + 1);

    ciSpearman = (exp(2*spearman_L) - 1) ./ (exp(2*spearman_L) + 1);


    %% ---------------------------------------------------------Test statistics

    df = nSizeP1 - 1;
    tValue = sampleR * sqrt(df/ (1 - sampleR^2));

    tProb = 2 * (1 - tcdf(abs(tValue), df));

    exactProb = rhocdf(sampleR, nSize-nPartialR, nullRho);



    %% ---------------------------------------------Combine like-minded outputs

    % All effect size point estimates
    esXECI = zeros(2,1);
    esXECI(1) = sampleR;
    esXECI(2) = unbiasedR;

    % All effect size interval estimates
    ciXECI = zeros(6,2);
    ciXECI(1,1:2) = ciSampleR;
    ciXECI(2,1:2) = ciUnbiased;
    ciXECI(3,1:2) = ciExact;
    ciXECI(4,1:2) = ciExactUnb;
    ciXECI(5,1:2) = ciPearson;
    ciXECI(6,1:2) = ciSpearman;

    % All relevant test statistic values
    tsXECI = zeros(4,1);
    tsXECI(1) = tValue;
    tsXECI(2) = df;
    tsXECI(3) = tProb;
    tsXECI(4) = exactProb;

    % All relevant variance estimates
    varXECI = zeros(2,1);
    varXECI(2) = seSampleR^2;



    %% ------------------------------------------If output display is requested

    if displayOut == 1
       
        strDecs = ['%10.', num2str(nDecs,0), 'f'] ;

        disp(blanks(3)');
        
        if nPartialR == 0
            str = [...
                ' Estimates of Effect Sizes for Pearson Correlation'; ...
                ' ================================================='];
        elseif nPartialR > 0
            str = [...
                ' Estimates of Effect Sizes for Partial Correlation'; ...
                ' ================================================='];
        end
        disp(str);
        disp(blanks(2)');

        label_esXECI = char(...
            '   Sample correlation:     ', ...
            '   Partial correlation:    ', ...
            '   Unbiased estimate:      ');

        label_ciXECI = char(...
            '   Large sample (r-to-z):  ', ...
            '   Large sample unbiased:  ', ...
            '   Exact:                  ', ...
            '   Exact unbiased:         ', ...
            '   Fisher-based Pearson:   ', ...
            '   Fisher-based Spearman:  ');

        label_tsXECI = char(...
            '   Observed t value:       ', ...
            '   Degrees of freedom:     ', ...
            '   P value:                ', ...
            '   P value [Rho <= r]:     ');

        nRows = length(ciXECI);
        c1 = char(40*ones(nRows,1));     % (
        c2 = char(44*ones(nRows,1));     % ,
        c3 = char(32*ones(nRows,1));     % space
        c4 = char(41*ones(nRows,1));     % )


        str = [...
            ' Point Estimates'; ...
            ' ---------------'];
        disp(str);
        if nPartialR == 0
            na = [1 3] ;
        else
            na = [2 3] ;
        end

        disp( horzcat( label_esXECI(na,:), num2str(esXECI, strDecs)) ) ;
        disp(blanks(1)');


        if (mod(ciSize,1) == 0)
            str = [' ' num2str(ciSize,'%6.0f') '% Confidence Intervals';' ------------------------'];
        else
            str = [' ' num2str(ciSize,'%6.1f') '% Confidence Intervals';' --------------------------'];
        end

        disp( horzcat(str));
        disp( horzcat( label_ciXECI, c1, num2str(ciXECI(:,1), strDecs), ...
                                 c2, c3, num2str(ciXECI(:,2), strDecs), c4) ) ;
        disp(blanks(1)');


        str = [...
            ' Test Statistics Results'; ...
            ' -----------------------'];
        disp(str);
        
        nc = [1:4] ;
        str_tsXECI = mixed2str(tsXECI(nc), [0 1 2 2], 10, nDecs) ;        

        disp( horzcat( label_tsXECI, str_tsXECI )) ;
        disp(blanks(1)') ;

    end
    %% ---------------------------------------------------------End of function

return

File: R functions/Correlation/XECI_corr.R
XECI_corr <-
  function(sampleR,
           nSize,
           nullRho = 0,
           ciSize = 95,
           partial = FALSE,
           nPartialR = 0) {
    # XECI_corr calculates effect sizes and CI for Pearson zero-order and
    # partial correlations. If Partial = TRUE, nPartial must be specified 
    # for meaningful results. 
    
    # Dependencies ####
    # Requires XECI_rho_cdf, XECI_ncp_rho, hypergeomF, is.wholeNumber
    
    # Argument specifications ####
    # sampleR    = observed correlation value
    # nSize      = sample size
    # ciSize     = confidence interval size
    # nullRho    = null hypothesised rho value
    # partial    = 1 if partial correlation provided, 0 if not 
    # nPartialR  = number of partialled variables
    
    # Output specifications ####
    # t test results
    # "Observed t value"= tVal
    # "Degrees of freedom" = df.tVal
    # "Obtained p value" = pVal.tTest
    # Point estimates
    # "partial correlation"= sampleR
    # "Unbiased Fisher\'s r to z" = unbiasedR
    # Confidence intervals
    # "Fisher\'s r to z CI" = fisher.RToZ.CI
    # "Unbiased Fisher\'s r to z CI" = unbiased.RToZ.CI
    # "Exact (sample) CI" = exactSample.CI
    # "Exact (unbiased) CI" = unbiasedExact.CI
    
    # VERSION HISTORY ####
    # 01.04.2018 - FST creates function - Only Fisher's r to z CI are calculated
    # 08.04.2018 - FST updates function with other functionality - does not account for partial correlations yet
    # 22.04.2018 - FST updates function with other functionality - accounts for partial correlations,  does not account for spearman correlation
    
    # Package from which functions were adapted for use in this function: ####
    #
    #
    
    # Input check ####
    # if partial does not equal 1, set npartialR to 0
    if(partial == FALSE){ nPartialR <- 0}
    # setting up a error message 
    stopCheck<-0
    # Check if sampleR is a number from -1 to 1
    if ((sampleR > 1) | (sampleR < -1) | !is.numeric(sampleR)) {
      warning("Sample r must be a number between -1 and +1, inclusive")
      stopCheck<-1
    }
    # Check if nSize is a whole number above 0 # REQUIRES is.wholeNumber function
    if ((nSize < 1) | !is.wholeNumber(nSize)) {
     warning("Sample size must be a whole number above 0")
      stopCheck<-1
    }
    # Check if CISize is a number between 1-99
    if ((ciSize <= 0) | (ciSize >= 100) | !is.numeric(ciSize)) {
      warning("Confidence interval width must be a number between 0 and 100, non-inclusive")
      stopCheck<-1
    }
    # Check if nullRho is a number between -1 to 1
    if ((nullRho > 1) | (nullRho < -1) | !is.numeric(nullRho)) {
      warning("Null Rho must be a number between -1 and +1, inclusive")
      stopCheck<-1
    }
    # Check if nPartialR is a whole number equal or greater than 0
    if (((nPartialR <0) | !is.wholeNumber(nPartialR))) {
      warning("The number of partialled variables must be a whole number above 0")
      stopCheck<-1
    }
        # stopping the program if there are any warnings
    if(stopCheck == 1) {
      stop("Input error")
      }
    
    # Calculations ####
    # setting up functions - fishers r to z and z to r 
    fisher.r2z <-
      function(sampleR) {
        0.5 * (log(1 + sampleR) - log(1 - sampleR))
      }
    fisher.z2r <- function(z) {
      ((exp(2 * z)) - 1) / ((exp(2 * z)) + 1)
    }
    #### preliminary calculations
    
    #calculating standard error
    SE <- 1 / sqrt(nSize- nPartialR - 3) 
    
    # P values for defining CI bounds
    probCI = c(((100+ciSize)/200), ((100-ciSize)/200))
    
    # calculating critical zs for rejection area
    zCritical <-  -qnorm(probCI)
    
    # t test
    # calculating t statistic
    tVal <- (sampleR * sqrt(nSize - 2)) / sqrt(1 - (sampleR ^ 2))
    # calculating df
    df.tVal <- nSize - 2
    # calculating p value
    pVal.tTest <- 2 * pt(tVal, df.tVal, lower.tail = FALSE)
    
    # Olkin and Pratt's (1958) unique minimum variance unbiased estimator of rho, uses Hypergeometric function of Pearson, J. W. (2009). 
    unbiasedR <- sampleR * hypergeomF(0.5, 0.5, (nSize-1-nPartialR)/2, 1 - sampleR^2)
  
    #### Confidence intervals ####
    
    # Fisher's r to z CI  for sample r ####
    #transforming r to z
    z <- fisher.r2z(sampleR)
    # calculating CI
    CI.Z <- z + zCritical * SE
    # Converting back to r
    fisher.RToZ.CI <- fisher.z2r(CI.Z)
    
    # Unbiased r CI using Fisher r to z CIs####
    #transforming r to z
    z.unbiasedR <- fisher.r2z(unbiasedR)
    # calculating CI
    CI.Z.UnbiasedR <-  z.unbiasedR + zCritical * SE
    # Converting back to r
    unbiased.RToZ.CI <- fisher.z2r(CI.Z.UnbiasedR)
    
    # Exact (sample) CI ####
    exactSample.CI <- 1
    exactSample.CI[1]  <- XECI_ncp_rho(sampleR, nSize-nPartialR, nPartialR, probCI[1])
    exactSample.CI[2]  <- XECI_ncp_rho(sampleR, nSize-nPartialR, nPartialR, probCI[2])
    # Exact (population) CI ####
    unbiasedExact.CI <- 1
    unbiasedExact.CI[1]  <- XECI_ncp_rho(unbiasedR, nSize-nPartialR, nPartialR, probCI[1])
    unbiasedExact.CI[2]  <- XECI_ncp_rho(unbiasedR, nSize-nPartialR, nPartialR, probCI[2])
    
    # output
    
    tTestOutput<-list("Observed t value" = tVal,
                      "Degrees of freedom" = df.tVal,
                      "Obtained p value" = pVal.tTest)
    pointEstimateOutput<-list("Sample correlation" = sampleR,
                              "Unbiased R" = unbiasedR)
    CIsOutput<-list("Fisher\'s r to z CI, sample r" = fisher.RToZ.CI,
                    "Fisher\'s r to z CI, unbiased r" = unbiased.RToZ.CI,
                    "Exact CI, sample r" = exactSample.CI,
                    "Exact CI, unbiased r" = unbiasedExact.CI)

    
    output <- list(tTestOutput,pointEstimateOutput,CIsOutput)
    return(output)
    
  #References####
    # Olkin, I., & Pratt, J. W. (1958). Unbiased Estimation of Certain Correlation Coefficients. The Annals of Mathematical Statistics, 29(1), 201-211.  Retrieved from http://www.jstor.org.ezp.lib.unimelb.edu.au/stable/2237306
    # Pearson, J. W. (2009). Computation of hypergeometric functions (Masters dissertation, University of Oxford).
  }